src/cml.js

```diff
@@ -0,0 +1,103 @@
+const { execSync } = require('child_process');
+const git_url_parse = require('git-url-parse');
+
+const Gitlab = require('./drivers/gitlab');
+const Github = require('./drivers/github');
+const { upload, exec } = require('./utils');
+
+const uri_no_trailing_slash = (uri) => {
+  return uri.endsWith('/') ? uri.substr(0, uri.length - 1) : uri;
+};
+
+const repo_from_origin = () => {
+  const origin = execSync('git config --get remote.origin.url').toString(
+    'utf8'
+  );
+  return git_url_parse(origin).toString('https').replace('.git', '');
+};
+
+const infer_driver = (opts = {}) => {
+  const { repo } = opts;
+  if (repo && repo.includes('github.com')) return 'github';
+  if (repo && repo.includes('gitlab.com')) return 'gitlab';
+
+  const { GITHUB_REPOSITORY, CI_PROJECT_URL } = process.env;
```

Comment:
What is none of the variables are defined?
Is it ok if it is defined but empty or wrong address (does it make sense to check URL pattern)?

Reply:
They are specific env variables for specific vendors. We are not even looking at the value we just need to 'see' them to 'know' which kind driver it is in case its not specified by the user